core/deployment/src/main/java/io/quarkus/deployment/util/WebJarUtil.java

```diff
@@ -0,0 +1,129 @@
+package io.quarkus.deployment.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+import io.quarkus.bootstrap.model.AppArtifact;
+import io.quarkus.bootstrap.model.AppDependency;
+import io.quarkus.deployment.pkg.builditem.CurateOutcomeBuildItem;
+
+/**
+ * Utility for Web resource related operations
+ */
+public class WebJarUtil {
+
+    private final static String tmpDir = System.getProperty("java.io.tmpdir");
+    private static final String OWN_MEDIA_FOLDER = "META-INF/resources/";
+    private static final List<String> IGNORE_LIST = Arrays.asList("logo.png", "favicon.ico");
+
+    private WebJarUtil() {
+    }
+
+    public static Path devOrTest(AppArtifact artifact, String rootFolderInJar)
+            throws IOException {
+
+        Path path = createDir(artifact.getGroupId(), artifact.getArtifactId(), artifact.getVersion());
+        if (isEmpty(path)) {
```

Comment:
What about multiple concurrenly running Quarkus instances? They might overwrite each other.
The safest thing would be to use `FileLock`, but this is not going to work for a directory, I fear...

Reply:
This is for dev mode, so we are saying that a developer will run concurrent versions of the same app at the exact same time ? I do not think we have to worry too much about this but I might be wrong.

Reply:
Ok I added a lock for the file when we create it.

Reply:
> This is for dev mode

And for tests, no?
Tests might be eve more prone to this. Imagine multiple concurrent builds on a Jenkins CI server (which doesn't run each build in a container).

> the same app

Doesn't need to be the same app. As long as those multiple apps use the same swagger-ui (etc.) version, a clash ist possible.

Btw, we are developing a (very) modular monolith and we in fact fire up multiple dev mode instances (in different submodules) in CI because they broke too often unnoticed.

Reply:
True. There is a lock in place now. Let's handle issues (if any) as we get them ?

Reply:
> Ok I added a lock for the file when we create it.

This is now done per file? Hm, how many files are there on average?
`FileLock` imposes some overhead so it would be good to limit it to one lock for all files but then again we already have the path check so I might be over-optimizing here...

Reply:
openapi = 8 files
graphql = 9 files
health = 9 files

I think it's ok like this ?

Reply:
> I think it's ok like this ?

Should be. You don't have any runtime numbers (in ms), do you?

Reply:
Nope